### PR TITLE
[JW8-11672] Reduce CLS when transitioning between standard and ad playback.

### PR DESCRIPTION
**Pull request by zetagame**
### This PR will...
Add a separate fullscreen button for IMA playback.

### Why is this Pull Request needed?
Changing the order of the buttons during playback was creating a costly CLS penalty.

### Are there any Pull Requests open in other repos which need to be merged with this?
Yes, see ima plugin repo.

#### Addresses Issue(s):

JW8-11672

### Checklist
- [ ] Jenkins builds and unit tests are passing
- [ ] I have reviewed the automated results

### CODE REVIEW

**Comment by jwrobot-player** [comment]
[:warning:](https://jenkins-player.longtailvideo.com/job/jw7-pr-multi-opensource/1755/) MULTI Build for commit 5d99dd1ab91d248a75185cecaeb4b31954ee13a1 is unstable (UNSTABLE).
[:building_construction:](https://jenkins-player.longtailvideo.com/job/jw7-opensource/9133/)[:black_small_square:](https://jenkins-player.longtailvideo.com/job/jw7-opensource/9133/consoleFull) jwplayer build SUCCESS
[:clipboard:](https://jenkins-player.longtailvideo.com/job/jw7-opensource-unit-tests/6122/)[:black_small_square:](https://jenkins-player.longtailvideo.com/job/jw7-opensource-unit-tests/6122/consoleFull) jwplayer unit tests SUCCESS
[:building_construction:](https://jenkins-player.longtailvideo.com/job/jw7-commercial/27926/)[:black_small_square:](https://jenkins-player.longtailvideo.com/job/jw7-commercial/27926/consoleFull) jwplayer-commercial build SUCCESS
[:clipboard:](https://jenkins-player.longtailvideo.com/job/jw7-commercial/23453/)[:black_small_square:](https://jenkins-player.longtailvideo.com/job/jw7-commercial-unit-tests/23453/consoleFull) jwplayer-commercial unit tests SUCCESS
[:cucumber: Cucumber Integration Tests](https://jenkins-player.longtailvideo.com/job/jw7-commercial-cucumber-wdio/5658/allure/) UNSTABLE
[:jack_o_lantern:](http://player-pr-test-jenkins.longtailvideo.com/builds/27926/archive/test/squash/) Manual Tests
[:tv:](http://player-pr-test-jenkins.longtailvideo.com/builds/27926/archive/test/view/) Views

File: src/js/view/controls/controlbar.js [hunk 1]
```diff
@@ -170,6 +170,9 @@ export default class Controlbar {
             cast: createCastButton(() => {
                 _api.castToggle();
             }, localization),
+            imaFullscreen: button('jw-icon-fullscreen', () => {
```
Comment (mamaddox) [inline src/js/view/controls/controlbar.js:173]: Nice!

File: src/css/controls/flags/ads.less [hunk 2]
```diff
@@ -53,9 +53,13 @@
...
         }
 
+        .jw-fullscreen-ads {
+            display: flex;
+        }
+
         .jw-tooltip,
         .jw-icon-tooltip:not(.jw-icon-volume),
-        .jw-icon-inline:not(.jw-icon-playback):not(.jw-icon-fullscreen):not(.jw-icon-volume) {
+        .jw-icon-inline:not(.jw-icon-playback):not(.jw-fullscreen-ads):not(.jw-icon-volume) {
```
Comment (mamaddox) [inline src/css/controls/flags/ads.less:62]: Is it okay that we will be changing the fullscreen placement for all ad types? Googima had to move to not clash with the IMA controls, but we usually have it in the lower right corner if we can. I'm cool with the change, just wondering.

File: src/js/view/controls/controlbar.js (hunk 1 above)
Comment (mamaddox) [reply src/js/view/controls/controlbar.js:173]: If this is the fullscreen icon for all ads do we want to call it something like `adsFullscreen`?

Comment (zetagame) [reply src/js/view/controls/controlbar.js:173]: good call. It was originally just for IMA and then I realized the shared UI. Forgot to update the name.

File: src/css/controls/flags/ads.less (hunk 2 above)
Comment (zetagame) [reply src/css/controls/flags/ads.less:62]: In all types we were using flex order to re-order it the same way. Now it's actually placed there to begin with

Comment (pajong) [reply src/css/controls/flags/ads.less:62]: Just trying to understand, are we now applying a new class `jw-fullscreen-ads` to IMA only skip button, and we still use the original `jw-icon-fullscreen` class for both the new and the old fullscreen? Change here seems to display `jw-fullscreen-ads` icon without displaying `jw-icon-fullscreen` as default - is that correct?

Comment (mamaddox) [reply src/css/controls/flags/ads.less:62]: I thought the flex order was just being used in IMA (which the related PR in IMA is removing). All the other ad types default to our control layout which has the icon in the right corner. Not against the change, just wanted to mention that.

**Comment by pajong** [comment]
@zetagame Has this been tested with FreeWheel/Vast/DAI plugins?

**Comment by zetagame** [comment]
> @zetagame Has this been tested with FreeWheel/Vast/DAI plugins?

Not fully. I'll do a spot check rn

File: src/js/view/controls/controlbar.js [hunk 3]
```diff
@@ -270,12 +278,15 @@ export default class Controlbar {
...
         _model.change('fullscreen', (model, val) => {
-            const fullscreenElement = this.elements.fullscreen.element();
-            toggleClass(fullscreenElement, 'jw-off', val);
-
-            const fullscreenText = model.get('fullscreen') ? localization.exitFullscreen : localization.fullscreen;
-            fullscreenTip.setText(fullscreenText);
-            setAttribute(fullscreenElement, 'aria-label', fullscreenText);
+            const fullscreenElements = [this.elements.fullscreen.element(), this.elements.adsFullscreen.element()];
+
+            for (let i=0; i < fullscreenElements.length; i++) {
```
Comment (pajong) [inline src/js/view/controls/controlbar.js:283]: Lint will fail - need space before/after `=`

Comment (zetagame) [reply src/js/view/controls/controlbar.js:283]: It actually passes lint:styles (maybe we should add a rule) but I'll change it for convention

**Comment by jwrobot-player** [comment]
[:warning:](https://jenkins-player.longtailvideo.com/job/jw7-pr-multi-opensource/1759/) MULTI Build for commit ede09da485438b51288a735d79e6b337f7ad48aa is unstable (UNSTABLE).
[:building_construction:](https://jenkins-player.longtailvideo.com/job/jw7-opensource/9137/)[:black_small_square:](https://jenkins-player.longtailvideo.com/job/jw7-opensource/9137/consoleFull) jwplayer build SUCCESS
[:clipboard:](https://jenkins-player.longtailvideo.com/job/jw7-opensource-unit-tests/6126/)[:black_small_square:](https://jenkins-player.longtailvideo.com/job/jw7-opensource-unit-tests/6126/consoleFull) jwplayer unit tests SUCCESS
[:building_construction:](https://jenkins-player.longtailvideo.com/job/jw7-commercial/27937/)[:black_small_square:](https://jenkins-player.longtailvideo.com/job/jw7-commercial/27937/consoleFull) jwplayer-commercial build SUCCESS
[:clipboard:](https://jenkins-player.longtailvideo.com/job/jw7-commercial/23462/)[:black_small_square:](https://jenkins-player.longtailvideo.com/job/jw7-commercial-unit-tests/23462/consoleFull) jwplayer-commercial unit tests SUCCESS
[:cucumber: Cucumber Integration Tests](https://jenkins-player.longtailvideo.com/job/jw7-commercial-cucumber-wdio/5669/allure/) UNSTABLE
[:jack_o_lantern:](http://player-pr-test-jenkins.longtailvideo.com/builds/27937/archive/test/squash/) Manual Tests
[:tv:](http://player-pr-test-jenkins.longtailvideo.com/builds/27937/archive/test/view/) Views

File: src/css/controls/flags/ads.less (hunk 2 above)
Comment (zetagame) [reply src/css/controls/flags/ads.less:62]: Ah you're right. It's just IMA

Comment (zetagame) [reply src/css/controls/flags/ads.less:62]: These kinds of insights are exactly why I posted this as a draft. I'll circle back and see if I can gate this changes to only IMA

**Review by pajong — APPROVED** [review]

**Comment by jwrobot-player** [comment]
[:warning:](https://jenkins-player.longtailvideo.com/job/jw7-pr-multi-opensource/1763/) MULTI Build for commit ffbdb46b4810c0c1df193639780389cd80b912e9 is unstable (UNSTABLE).
[:building_construction:](https://jenkins-player.longtailvideo.com/job/jw7-opensource/9141/)[:black_small_square:](https://jenkins-player.longtailvideo.com/job/jw7-opensource/9141/consoleFull) jwplayer build SUCCESS
[:clipboard:](https://jenkins-player.longtailvideo.com/job/jw7-opensource-unit-tests/6130/)[:black_small_square:](https://jenkins-player.longtailvideo.com/job/jw7-opensource-unit-tests/6130/consoleFull) jwplayer unit tests SUCCESS
[:building_construction:](https://jenkins-player.longtailvideo.com/job/jw7-commercial/27946/)[:black_small_square:](https://jenkins-player.longtailvideo.com/job/jw7-commercial/27946/consoleFull) jwplayer-commercial build SUCCESS
[:clipboard:](https://jenkins-player.longtailvideo.com/job/jw7-commercial/23469/)[:black_small_square:](https://jenkins-player.longtailvideo.com/job/jw7-commercial-unit-tests/23469/consoleFull) jwplayer-commercial unit tests SUCCESS
[:cucumber: Cucumber Integration Tests](https://jenkins-player.longtailvideo.com/job/jw7-commercial-cucumber-wdio/5677/allure/) UNSTABLE
[:jack_o_lantern:](http://player-pr-test-jenkins.longtailvideo.com/builds/27946/archive/test/squash/) Manual Tests
[:tv:](http://player-pr-test-jenkins.longtailvideo.com/builds/27946/archive/test/view/) Views

**Review by mamaddox — APPROVED** [review]

**Pull request merged**